### ISSUE TITLE
`colcon build` uses the wrong Python version when inside a virtualenvironment on Windows

### ISSUE THREAD

**Issue by RobertoRoos** [colcon/colcon-core#598]
See also this stackexchange issue: https://robotics.stackexchange.com/questions/105587 

I have ROS2 running from a venv. And when building a workspace I get a bunch of ModuleNotFound errors, because colcon is using the wrong Python executable according to the logs.  
It's using `C:\Program Files\Python310\python.exe` instead of `C:\dev\ros2_humble\.venv\Scripts\python.exe`.

So colcon does not use the currently activated virtualenvironment, which seems wrong.

On Windows 10, with ROS2 humble, and Python 3.8 (used as venv).

**Comment by RobertoRoos** [comment]
Actually, I doubt this is the fault of colcon.

In a fresh directory I created the following `CMakeLists.txt`:

```
cmake_minimum_required(VERSION 3.5)
project("PythonDemo")

find_package(Python COMPONENTS Interpreter Development)

message(STATUS "Python_EXECUTABLE: ${Python_EXECUTABLE}")
```

And build from inside ROS2:

```
> C:\dev\ros2_humble\.venv\Scritps\Activate.ps1
> C:\dev\ros2_humble\local_setup.ps1
> mkdir build; cd build
> python -V

Python 3.8.10

> cmake ..

-- Python_EXECUTABLE: C:/Program Files/Python310/python.exe
```

It seems CMake who doesn't pick up the right version.

**Comment by andanders** [comment]
I just ran into this problem. Is there a best practice for selecting the right python binary? In my system cmake selects the binary in my anaconda dir even though I disable anaconda for the bash session.

**Comment by cottsay** [comment]
There are a few factors at play here.
1. When building `python` or `ros.ament_python` packages, colcon will always use the same interpreter as it was installed with. So if you install colcon using debs, it will be the system's Python interpreter regardless of what venvs or anaconda configurations you have enabled. You can work around this by either:
   * Invoking colcon explicitly using the interpreter you'd like to use, i.e. `python3 -m colcon build ...` or `/usr/local/bin/python -m colcon build ...`
   * Installing a new copy of `colcon-core` using your new interpreter, possibly using pip. If your new interpreter doesn't see system packages (see the `--system-site-packages` option to venv), you'd need to install all of the colcon packages and not just `colcon-core`. In practice, this is actually true of pretty much all of the ROS 2 python package dependencies, which is a significant part of why we don't recommend using venvs or anaconda.
2. CMake will determine which interpreter to use on its own, completely independent of what colcon is using. The default behavior is to find the "newest" version of Python, which is not necessarily the one that gets invoked when you run the `python` or `python3` executable.
3. We [tried to work around](https://github.com/ament/ament_cmake/pull/507) (2) a little bit in ROS 2 humble by adding a hint, so the situation in ROS 2 has changed a little in recent years.

With so many technologies at play here and without a consistent and reliable way to make them all behave the same, the recommendation we've been making for years remains the same: Just use your system's default Python interpreter, and don't use venvs. It's absolutely possible to make all of this work together, but you should expect to see problems like this if you do. It's simply not supported.

**Comment by andanders** [comment]
To your first point "colcon will always use the same interpreter as it was installed with", it is my understanding that colcon is very picky about the version of python. If ROS2/colcon was build with python 3.12, then it will only run with that same version.
In my case the default interpreter is Python 3.11, and that caused problems right off the bat. Would it be appropriate for me to create an issue with colcon, that it should raise an error or at least a warning when being installed with a default interpreter that is incompatible because it is a different version?

**Comment by gavanderhoorn** [comment]
> it is my understanding that colcon is very picky about the version of python. If ROS2/colcon was build with python 3.12, then it will only run with that same version.

in my experience Colcon doesn't really care, as long as its Python packages are available to the interpreter you're trying to use (ie: it has to have been installed for it).

There are no native (ie: compiled) parts to Colcon (at least not in the default extensions), so as long as the Python interpreter can execute the scripts it should work.

What is tied to a specific version of Python though are the native parts of ROS 2 that packages like `rclpy` *wrap*. Binary compatibility must be guaranteed there and that's where installing those binary parts 'for' one version of Python and trying to use them with another won't work.

(but that (ie: ABI compatibility) is always an issue, not unique to ROS 2)

**Comment by cottsay** [comment]
To add to @gavanderhoorn's great explanation:

We test colcon with Python 3.6 up to Python 3.12 today. I have 3.13 testing on the roadmap but the removal of `lib2to3` broke our spelling linters so it's still not running right now, but I run Python 3.13 on my development machine and can say that I'm not aware of any problems.

tl;dr - as far as I am aware, colcon itself should behave the same on every currently supported Python version, full stop.

**Comment by andanders** [comment]
Ah yes, I was not very accurate. What gavanderhoorn said hit the mark. "_... parts of ROS 2 that packages like rclpy wrap..._" so that is not colcons faultt.
